core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -119,16 +156,25 @@ private void runThreadLocalContextCustomizer() {
         }
     }
 
-    private DefaultClientRequestContext(DefaultClientRequestContext ctx) {
-        this(ctx, ctx.request());
+    private void failEarly(Exception cause) {
+        final RequestLogBuilder logBuilder = logBuilder();
+        final UnprocessedRequestException wrapped = new UnprocessedRequestException(cause);
+        logBuilder.endRequest(wrapped);
+        logBuilder.endResponse(wrapped);
+
+        final Request req = request();
+        if (req instanceof StreamMessage) {
+            ((StreamMessage<?>) req).abort();
+        }
     }
 
-    private DefaultClientRequestContext(DefaultClientRequestContext ctx, Request request) {
+    private DefaultClientRequestContext(DefaultClientRequestContext ctx, Request request, Endpoint endpoint) {
         super(ctx.meterRegistry(), ctx.sessionProtocol(), ctx.method(), ctx.path(), ctx.query(), request);
 
         eventLoop = ctx.eventLoop();
         options = ctx.options();
-        endpoint = ctx.endpoint();
+        endpointSelector = ctx.endpointSelector();
+        this.endpoint = requireNonNull(endpoint, "endpoint");
```

Comment:
Very minor nit but this looks a little out of place and maybe could go after setting all the `ctx` fields.

Reply:
They are all `ctx` fields (`eventLoop`, `options`, `endpointSelector` and `this.endpoint`).